streams/shared/src/main/scala/zio/stream/StreamEffect.scala

```diff
@@ -17,22 +17,33 @@
 package zio.stream
 
 import zio._
-
-private[stream] class StreamEffect[+E, +A](val processEffect: Managed[E, () => A])
-    extends ZStream[Any, E, A](
-      processEffect.map { thunk =>
-        UIO.effectTotal {
-          try UIO.succeed(thunk())
-          catch {
-            case StreamEffect.Failure(e) => IO.fail(Some(e.asInstanceOf[E]))
-            case StreamEffect.End        => IO.fail(None)
-          }
-        }.flatten
+import zio.internal.Executor
+
+private[stream] class StreamEffect[-R, +E, +A](
+  val processEffect: ZManaged[R, E, () => A],
+  val executor: ZIO[R, Nothing, Executor]
```

Comment:
As far as I can tell, this structure doesn't compose with well-defined semantics. Consider, e.g. `x ++ y` — what is the `executor` for such a composite?

It seems like we need to say three things:

 - Invocation of "open" occurs on the blocking thread pool, which we can do in the `reserve` of the managed
 - Invocation of "close" occurs on the blocking thread pool, which we can do in the `reserve` of managed (on finalization)
 - Invocation of the "pull" occurs on the blocking thread pool, which we can also do (below, using `effectBlocking` instead of `effectTotal`, etc.).

So unless I'm missing something, we can do this w/o new machinery. Right? Or wrong? 😄

Reply:
Say `x: StreamEffect` uses a Blocking executor, `y: StreamEffect` uses a regular executor, and `z: Stream` that is asynchronous.

`x ++ y ++ z` does the following:
- opens `x` and starts pulling it. The `Pull` on `x` evaluates the `() => A` thunk while locked to the Blocking executor, so that pull runs on the blocking executor;
- opens `y` and starts pulling it. Similarly, it evaluates the thunk while locked to the default executor, so that pull runs there.
- opens `z` and starts pulling it. That `Pull` does whatever it wants because it's effectful.

From how I'm reading things, the semantics here are identical to the semantics of `lock`, as long as we always propagate the executor. Wdyt?

Reply:
@jdegoes 

Sorry, I realized I didn't respond to your questions about whether it's possible to do this without new machinery.

The problem is as follows: how does a constructor (like `fromInputStream`), which only produces a thunk `Managed[E, () => A]`, indicate that when a stream which is based on it is run, it should be run on the Blocking threadpool? We can't use `effectBlocking` because we want to produce synchronous code that runs in the thunk.

Reply:
@jdegoes Any thoughts on the above? :-)

Reply:
We hashed it over deadlift sets. I believe @jdegoes is onboard. I'll pick this back up next week :-)

Reply:
Awesome. Please keep me in the review loop.